rocksdb_replicator/replicated_db.cpp

```diff
@@ -139,11 +146,32 @@ rocksdb::Status RocksDBReplicator::ReplicatedDB::Write(
     // TODO(bol): This potentially could block all worker threads. We may
     // consider having a dedicated set of worker threads for admin requests,
     // and/or provide async write API when this turns out to be a problem.
-    if (!max_seq_no_acked_.wait(cur_seq_no, FLAGS_replicator_timeout_ms)) {
+    if (!max_seq_no_acked_.wait(cur_seq_no, current_replication_timeout_ms_)) {
       incCounter(kReplicatorWriteWaitTimedOut, 1, db_name_);
       LOG(ERROR) << "Failed to receive ack from follower, timing out for " << db_name_;
+      numConsecutiveReplTimeout_ ++;
+
+      // enter degradation mode if there has been consecutive timeouts waiting for the follower ACK.
+      // This allows us to use a smaller wait timeout to fail fast, instead of blocking for long.
+      if (current_replication_timeout_ms_ == FLAGS_replicator_timeout_ms
+          && numConsecutiveReplTimeout_ == kNumReplTimeoutsBeforeDegradation
+          && FLAGS_replicator_timeout_degraded_ms < FLAGS_replicator_timeout_ms
+          && FLAGS_replicator_timeout_degraded_ms >= kMinReplTimeoutMs) {
+        current_replication_timeout_ms_ = FLAGS_replicator_timeout_degraded_ms;
```

Comment:
To clarify, this degraded state is per db (shard), right? If one shard is degraded, the other shards continue to operate as usual without a change in timeouts. Is that accurate?

Reply:
yeah, the timeout `current_replication_timeout_ms_ ` is stored per db, also added tests for this

Reply:
Also verified in the introspection page:

<img width="439" alt="image" src="https://user-images.githubusercontent.com/9246439/163264617-720fb3a4-5868-4699-a967-f24ae03b0c74.png">